File: src/test/java/com/digitalsanctuary/spring/demo/user/ui/page/UpdateUserPage.java
package com.digitalsanctuary.spring.demo.user.ui.page;

import static com.codeborne.selenide.Condition.appear;
import static com.codeborne.selenide.Condition.cssClass;
import static com.codeborne.selenide.Condition.visible;
import com.codeborne.selenide.Selenide;
import com.codeborne.selenide.SelenideElement;

/**
 * Page object for update user profile page
 */
public class UpdateUserPage {

    private final SelenideElement FIRST_NAME_FIELD = Selenide.$x("//input[@id='firstName']");
    private final SelenideElement LAST_NAME_FIELD = Selenide.$x("//input[@id='lastName']");
    private final SelenideElement UPDATE_BUTTON = Selenide.$x("//button[@type='submit']");
    private final SelenideElement GLOBAL_MESSAGE = Selenide.$x("//div[@id='globalMessage']");
    private final SelenideElement CHANGE_PASSWORD_LINK = Selenide.$x("//a[@href='/user/update-password.html']");
    private final SelenideElement DELETE_ACCOUNT_LINK = Selenide.$x("//a[@href='/user/delete-account.html']");

    public UpdateUserPage(String url) {
        Selenide.open(url);
    }

    public UpdateUserPage() {
        // For navigation to this page from other pages
    }

    /**
     * Update user profile information
     */
    public UpdateUserPage updateProfile(String firstName, String lastName) {
        FIRST_NAME_FIELD.clear();
        FIRST_NAME_FIELD.setValue(firstName);
        LAST_NAME_FIELD.clear();
        LAST_NAME_FIELD.setValue(lastName);
        UPDATE_BUTTON.click();

        // Wait for success message to appear
        GLOBAL_MESSAGE.should(appear);
        return this;
    }

    /**
     * Get the success/error message displayed after update
     */
    public String getMessage() {
        return GLOBAL_MESSAGE.shouldBe(visible).text();
    }

    /**
     * Check if update was successful
     */
    public boolean isUpdateSuccessful() {
        return GLOBAL_MESSAGE.has(cssClass("alert-success"));
    }

    /**
     * Navigate to change password page
     */
    public UpdatePasswordPage goToChangePassword() {
        CHANGE_PASSWORD_LINK.click();
        return new UpdatePasswordPage();
    }

    /**
     * Navigate to delete account page
     */
    public DeleteAccountPage goToDeleteAccount() {
        DELETE_ACCOUNT_LINK.click();
        return new DeleteAccountPage();
    }

    /**
     * Get current first name value
     */
    public String getFirstName() {
        return FIRST_NAME_FIELD.getValue();
    }

    /**
     * Get current last name value
     */
    public String getLastName() {
        return LAST_NAME_FIELD.getValue();
    }

    /**
     * Wait for page to load completely
     */
    public UpdateUserPage waitForPageLoad() {
        FIRST_NAME_FIELD.should(appear);
        LAST_NAME_FIELD.should(appear);
        UPDATE_BUTTON.should(appear);
        return this;
    }
}

File: playwright/src/pages/UpdateUserPage.ts
import { Page, Locator, expect } from '@playwright/test';
import { BasePage } from './BasePage';

/**
 * Page Object for the Update User Profile page.
 */
export class UpdateUserPage extends BasePage {
  readonly path = '/user/update-user.html';

  // Form elements
  readonly firstNameInput: Locator;
  readonly lastNameInput: Locator;
  readonly submitButton: Locator;

  // Messages
  readonly globalMessage: Locator;

  // Error displays
  readonly firstNameError: Locator;
  readonly lastNameError: Locator;

  // Navigation links
  readonly changePasswordLink: Locator;
  readonly deleteAccountLink: Locator;

  constructor(page: Page) {
    super(page);
    this.firstNameInput = page.locator('#firstName');
    this.lastNameInput = page.locator('#lastName');
    // Use specific button text to avoid matching logout button
    this.submitButton = page.getByRole('button', { name: 'Update Profile' });
    this.globalMessage = page.locator('#globalMessage');
    this.firstNameError = page.locator('#firstNameError');
    this.lastNameError = page.locator('#lastNameError');
    // Use role-based selectors with exact name to get visible buttons/links
    this.changePasswordLink = page.getByRole('link', { name: 'Change Password' }).locator('visible=true').first();
    this.deleteAccountLink = page.getByRole('link', { name: 'Delete Account' }).locator('visible=true').first();
  }

  /**
   * Get current first name value.
   */
  async getFirstName(): Promise<string> {
    return this.firstNameInput.inputValue();
  }

  /**
   * Get current last name value.
   */
  async getLastName(): Promise<string> {
    return this.lastNameInput.inputValue();
  }

  /**
   * Fill in profile form.
   */
  async fillForm(firstName: string, lastName: string): Promise<void> {
    await this.firstNameInput.clear();
    await this.firstNameInput.fill(firstName);
    await this.lastNameInput.clear();
    await this.lastNameInput.fill(lastName);
  }

  /**
   * Submit the form.
   */
  async submit(): Promise<void> {
    await this.submitButton.click();
  }

  /**
   * Update profile.
   */
  async updateProfile(firstName: string, lastName: string): Promise<void> {
    await this.fillForm(firstName, lastName);
    await this.submit();
  }

  /**
   * Update profile and wait for success message.
   */
  async updateProfileAndWait(firstName: string, lastName: string): Promise<void> {
    await this.updateProfile(firstName, lastName);
    await this.waitForMessage();
  }

  /**
   * Wait for message to appear.
   */
  async waitForMessage(timeout = 5000): Promise<void> {
    await this.globalMessage.waitFor({ state: 'visible', timeout });
  }

  /**
   * Get message text.
   */
  async getMessageText(): Promise<string | null> {
    if (await this.globalMessage.isVisible()) {
      return this.globalMessage.textContent();
    }
    return null;
  }

  /**
   * Check if message indicates success.
   */
  async isSuccessMessage(): Promise<boolean> {
    const classes = await this.globalMessage.getAttribute('class');
    return classes?.includes('alert-success') || classes?.includes('alert-info') || false;
  }

  /**
   * Navigate to change password page.
   */
  async goToChangePassword(): Promise<void> {
    await this.changePasswordLink.click();
    await this.page.waitForURL('**/update-password**');
  }

  /**
   * Navigate to delete account page.
   */
  async goToDeleteAccount(): Promise<void> {
    await this.deleteAccountLink.click();
    await this.page.waitForURL('**/delete-account**');
  }
}
